### PR TITLE
Expressive empty states

### PR DESCRIPTION
**Pull request by msecret**
Adds empty states for:
- no orgs
- no users
- just one user
- no spaces

Modifies the existing no apps state for pages like overview where there are many side by side.

Still needs services, which is a larger one and not required for usability testing, so don't need for this iteration.

Merged in #965 
Requires https://github.com/18F/cg-style/pull/313
Ref https://favro.com/card/1e11108a2da81e3bd7153a7a/18F-556

image

image

### CODE REVIEW

File: static_src/components/app_list.jsx [hunk 1]
```diff
@@ -65,7 +69,11 @@ export default class AppList extends React.Component {
     );
 
     if (this.state.empty) {
-      content = <h4 className="test-none_message">No apps</h4>;
+      content = (
+        <EntityEmpty callout="You have no apps in this space">
+          <InfoAppCreate space={ this.state.currentSpace } org={ this.state.currentOrg } />
```
Comment (adborden) [inline static_src/components/app_list.jsx]: Do we want to check the user roles here?

File: static_src/css/main.css [hunk 2]
```diff
@@ -36,3 +36,4 @@ input[type="checkbox"] + label::before,
 input[type="radio"] + label::before {
   display: none;
 }
+
```
Comment (adborden) [inline static_src/css/main.css]: Hmm, I didn't even know we had CSS in here...

File: static_src/stores/user_store.js [hunk 3]
```diff
@@ -249,19 +310,20 @@ class UserStore extends BaseStore {
...
+  get isLoadingCurrentUser() {
+    return this._loading.currentUser === true;
   }
 
-  currentUserHasSpaceRole(role) {
-    return this._hasRole(role, 'space_roles');
-  }
+  hasRole(userGuid, entityGuid, roleToCheck) {
+    let wrappedRoles = roleToCheck;
+    if (!roleToCheck.length) {
```
Comment (adborden) [inline static_src/stores/user_store.js]: We should check the type here, both strings and arrays have length.

File: static_src/stores/user_store.js [hunk 4]
```diff
@@ -249,19 +310,20 @@ class UserStore extends BaseStore {
...
+      wrappedRoles = [roleToCheck];
+    }
 
-  currentUserHasOrgRole(role) {
-    return this._hasRole(role, 'organization_roles');
+    const key = entityGuid;
+    const user = this.get(userGuid);
+    const roles = user && user.roles && user.roles[key] || [];
+    return !!roles.find((role) => wrappedRoles.includes(role));
```
Comment (adborden) [inline static_src/stores/user_store.js]: We should add documentation for this method, I was thinking if you passed an array of roles, the check would only succeed if the user has ALL roles, but this is checking for ANY role which is a big difference.

File: static_src/test/server/api.js [hunk 5]
```diff
@@ -19,9 +19,17 @@ var spaceRoutes = require('./fixtures/space_routes');
...
+var userOrganizations = require('./fixtures/user_organizations.js');
+var userSpaces = require('./fixtures/user_spaces.js');
 
 var BASE_URL = '/v2';
 
+const ENV_NO_ORGS = process.env.NO_ORGS || false;
+const ENV_NO_SPACES = process.env.NO_SPACES || false;
+const ENV_NO_APPS = process.env.NO_APPS || false;
+const ENV_NO_ORG_USERS = process.env.NO_ORG_USERS || false;
+const ENV_NO_SPACE_USERS = process.env.NO_SPACE_USERS || false;
```
Comment (adborden) [inline static_src/test/server/api.js]: Hmm, I wonder if we can use the smocks admin to pre-program this. Not sure how smocks persists its data...

**Comment by adborden** [comment]
We should be writing more component tests too, I'll add one for `InfoAppCreate`

File: package.json [hunk 6]
```diff
@@ -18,7 +18,7 @@
     "test-unit": "NODE_ENV=test karma start --single-run",
     "test-selenium-install": "selenium-standalone install",
     "testing-server": "PORT=8001 npm run testing-server-run",
-    "testing-server-run": "nodemon --watch ./static_src/test/server ./static_src/test/server/server.js",
+    "testing-server-run": "node ./static_src/test/server/server.js",
```
Comment (adborden) [inline package.json]: Don't forget to remove it from `package.json`

File: static_src/actions/user_actions.js [hunk 7]
```diff
@@ -122,11 +125,172 @@ export default {
...
+  receivedUser(user) {
+    AppDispatcher.handleServerAction({
+      type: userActionTypes.USER_RECEIVED,
+      user
+    });
+
+    return Promise.resolve(user);
+  },
+
+  fetchUserSummary(userGuid) {
```
Comment (msecret) [inline static_src/actions/user_actions.js]: I may have left it in.

File: static_src/components/app_list.jsx (hunk 1 above)
Comment (msecret) [inline static_src/components/app_list.jsx]: I think I moved the roles to `InfoAppCreate`.

File: static_src/css/main.css (hunk 2 above)
Comment (msecret) [inline static_src/css/main.css]: Yea we were using it for quick styles at first, but found that putting it in cg-style was easier.

File: static_src/stores/user_store.js (hunk 3 above)
Comment (msecret) [inline static_src/stores/user_store.js]: Keep forgetting that.

File: static_src/test/server/api.js (hunk 5 above)
Comment (msecret) [inline static_src/test/server/api.js]: eya I keep forgetting about that admin

File: static_src/components/app_list.jsx (hunk 1 above)
Comment (adborden) [inline static_src/components/app_list.jsx]: Yeah, I think you still need to pass in a user to `props` for it to do the check.

**Pull request merged**